File: hard/print_check.rb
$s = ["", "One", "Two", "Three", "Four",
     "Five", "Six", "Seven", "Eight", "Nine",
     "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen",
     "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen",
     "Twenty", "Thirty", "Forty", "Fifty",
     "Sixty", "Seventy", "Eighty", "Ninety",
     "Hundred", "Thousand", "Million"]

def wrd(a, b, c)
  return false if a + b + c == 0
  print $s[a] + $s[28] if a > 0
  if b > 1
    print $s[18 + b] + $s[c]
  else
    print $s[10*b + c]
  end
  true
end

File.open(ARGV[0]).each_line do |line|
  if line.to_i == 0
    print "Zero"
  else
    l = line.chomp.split("").map(&:to_i)
    while l.length < 9 do l.insert(0, 0) end
    print $s[30] if wrd(l[0], l[1], l[2])
    print $s[29] if wrd(l[3], l[4], l[5])
    wrd(l[6], l[7], l[8])
  end
  puts "Dollars"
end

File: hard/print_check.lua
function div(a, b) return math.floor(a/b) end

s = {[0]="", "One", "Two", "Three", "Four",
     "Five", "Six", "Seven", "Eight", "Nine",
     "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen",
     "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen",
     "Twenty", "Thirty", "Forty", "Fifty",
     "Sixty", "Seventy", "Eighty", "Ninety",
     "Hundred", "Thousand", "Million"}

function wrd(a1, a2, a3)
  if a1 + a2 + a3 == 0 then return false end
  if a1 > 0 then
    io.write(s[a1] .. s[28])
  end
  if a2 > 1 then
    io.write(s[18 + a2] .. s[a3])
  else
    io.write(s[10*a2 + a3])
  end
  return true
end

for line in io.lines(arg[1]) do
  local b, c = tonumber(line), {}
  if b == 0 then
    io.write("Zero")
  else
    for i = 9, 1, -1 do
      c[i], b = b%10, div(b, 10)
    end
    if wrd(c[1], c[2], c[3]) then io.write(s[30]) end
    if wrd(c[4], c[5], c[6]) then io.write(s[29]) end
    _ = wrd(c[7], c[8], c[9])
  end
  io.write("Dollars\n")
end
